src/core/lib/iomgr/ev_poll_posix.cc

```diff
@@ -1439,7 +1446,30 @@ static void cache_destroy_locked(poll_args* args) {
     poll_cache.free_pollers = args->next;
   }
 
-  gpr_free(args);
+  // Now move this args to the dead poller list for later join
+  if (poll_cache.dead_pollers != nullptr) {
+    poll_cache.dead_pollers->prev = args;
+  }
+  args->prev = nullptr;
+  args->next = poll_cache.dead_pollers;
+  poll_cache.dead_pollers = args;
+}
+
+static void cache_harvest_locked() {
+  while (poll_cache.dead_pollers) {
+    poll_args* args = poll_cache.dead_pollers;
+    poll_cache.dead_pollers = poll_cache.dead_pollers->next;
+    // Keep the list consistent in case new dead pollers get added when we
+    // release the lock below to wait on joining
+    if (poll_cache.dead_pollers) {
+      poll_cache.dead_pollers->prev = nullptr;
+    }
+    gpr_cv_signal(&args->harvest);
+    gpr_cv_wait(&args->join, &g_cvfds.mu, gpr_inf_future(GPR_CLOCK_MONOTONIC));
+    args->poller_thd.Join();
+    args->poller_thd.~Thread();
```

Comment:
I have seen this pattern  of `Join` followed by `~Thread()` in many places. Would it help to add a parameter to `Join` so that it calls `~Thread()`  instead of making two explicit calls ? 

A tangential point:
Somehow calling ~Thread() explicitly feels a bit inconvenient to me :) ..I have rarely seen this pattern. Also from a readability perspective (of someone like me who is not super familiar with C++) ,  calling `~Foo()` mentally registers as the memory of the object of type `Foo` as also being destroyed; but that is not the case here; We are merely invoking the destructor (and not freeing any memory. The actual freeing of memory happens when the structure containing this thd is destroyed).. 

So might I suggest creating another helper function called like `TearDown` or `Finish` or something like that and put whatever code you have in `~Thead()` in that function?  (Then `~Thead()` (and perhaps `Join()` can simply call `TearDown()`).

Reply:
So right now we use manual destruction only in cases where we're a member of a struct that is allocated by `gpr_malloc` and released by `gpr_free`. Once these move to a real class allocated by `New` and freed by `Delete`, the actual destructor will just get called.